### DISCUSSION TITLE
請問Python中 max() 和 maximum() 的差別

### DISCUSSION THREAD

**Discussion by Yunhans** [1101-GS3073-C/Python-Community-GS3073-C#29] [Q&A] [answered]
如果我只用max()如下

```
from pylab import *

x = linspace(0,10*pi,100)
y = cos(x)**2 / sqrt(max(1,2*x-1))

plot(x,y)

show()
```

會跑出ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
但是如果我用了maximum()

```
from pylab import *

x = linspace(0,10*pi,100)
y = cos(x)**2 / sqrt(maximum(1,2*x-1))

plot(x,y)

show()
```

就能成功跑出函式圖
請問兩者差別

**Comment by briankuo123** [comment, marked as answer]
我認為問題應該是出在max()無法比較單一數字跟array的大小，所以才會跑出這個錯誤。
max() 用法應該是 max(x ,y, z) 或是 max(array) 這樣。
而 maximum() 的用法本身就是 maximum(array1, array2) 所以就可以順利跑出函式圖。
這個兩個array比較可以是1對多或是兩個相同元素量的陣列做比較

**Comment by chia-kai-chang** [comment]
謝謝 @briankuo123 的回應

我來補充一下

### max() 輸出為單一element
```
print(max(1,2,3))
# 3
```

### maximum()輸出為ndarray
```
import numpy as np
print(maximum(np.array([1,2,3]), np.array([4,5,1])))
# [4 5 3]
```

<img width="548" alt="PNG影像" src="https://user-images.githubusercontent.com/1768891/136047194-285c235c-1b65-42ce-82d8-6a4399184004.png">


@Yunhans 你的程式中，`linspace` 的結果是numpy的ndarray。這點要注意。

**Comment by jim11261314** [comment]
對初學者來說array相關的函式都是痛苦的斷層
老師與同學講解得非常清楚
